### DISCUSSION TITLE
Shake-256 for signing

### DISCUSSION THREAD

**Discussion by DarkBlackJPG** [randombit/botan#3026] [Q&A] [answered]
Hey guys, I'm making a simple project for my masters thesis regarding signature schemes and I need to sign a document using SHA-256 and SHAKE-256, I can't seem to find any way to use SHAKE-256 as a hash function for document/file signing.

I'm guessing you can theoretically use that algorithm for signing, correct?

If there's no internal implementation, does botan provide some kind of API for creating my own "signature scheme" with any hash function?

I've currently tried all EMSA1(SHAKE-256), EMSA2(SHAKE-256), EMSA3(SHAKE-256), EMSA4(SHAKE-256) when creating the PK_signer object.

This is a bit over my head so be gentle hahah
Thanks to everyone in advance :smiley:

**Comment by reneme** [comment, marked as answer]
Traditionally, signing is done with asymmetric schemes where you have a private key (performing the sign operation) and an associated public key (used in the verify information). This is done with asymmetric algorithms like RSA, ECDSA and the like.

"Signing a document" with just a hash function (i.e. SHA-256 or SHAKE-256) doesn't really make sense cryptographically, as this does not involve a "secret", let alone a public/private key pair. Think of hash functions as producing a "fingerprint" of your document that could be easily computed by anyone. Whereas a valid signature can only be calculated by the person holding the private key.

Now: Do you want to actually obtain a publicly verifiable signature of your document (as explained in the first paragraph), or do you want to "just" hash the document?

**Reply by DarkBlackJPG** [reply]
Hey, sorry if I wasn't precise. 
_TL;DR: Essentially, I want to obtain a publicly verifiable signature_

I want to use RSA-2048 with SHAKE-256 & SHA-256. The hash would be used, like you said, only as a fingerprint. However, I can't seem to figure out how configure my app to work with SHAKE.
Here's the sample code for signing:
```
try {
        Botan::PK_Signer pkSigner(*rsaPrivateKey, rng, emsa);
        pkSigner.update(data);
        std::vector<uint8_t> signature = pkSigner.signature(rng);
#ifdef LOGGING_ENABLED
        spdlog::info("Byte array data was signed.");
#endif
        return signature;
```
Where the emsa value is currently configured as `EMSA1(SHA-256)`.

**Reply by reneme** [reply]
Shake has a variable output length that must be specified. Please try `EMSA1(SHAKE-256(256))`.

**Reply by reneme** [reply]
And on a second note: Since you appear to use an RSA private key, you should consider using a different padding scheme. E.g. Instead of `EMSA1(SHA-256)` use `PSSR(SHA-256,MGF1,32)`.

**Reply by DarkBlackJPG** [reply]
That's it, thanks!
I didn't know I had to define the output length 
Also thanks for the heads up on the padding scheme :smile:
